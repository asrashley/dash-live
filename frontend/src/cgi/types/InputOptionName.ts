export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}