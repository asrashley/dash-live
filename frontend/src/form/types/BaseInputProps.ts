import type { ReadonlySignal } from "@preact/signals";
import { StaticInputProps } from "./StaticInputProps";

export type BaseInputProps = {
  name: string;
  id: string;
  type: StaticInputProps["type"];
  className: string | ReadonlySignal<string>;
  title: string;
  placeholder?: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

