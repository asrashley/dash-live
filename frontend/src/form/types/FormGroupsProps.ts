import { InputFormGroup } from "./InputFormGroup";
import { FormRowMode } from "./FormRowMode";
import { SetValueFunc } from "./SetValueFunc";
import { InputProps } from "./InputProps";

export interface FormGroupsProps extends Pick<InputProps, "data" | "disabledFields"> {
  groups: InputFormGroup[];
  expand?: string;
  mode: FormRowMode;
  layout?: number[];
  setValue: SetValueFunc;
}
