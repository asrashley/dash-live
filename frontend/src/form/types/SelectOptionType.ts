export interface SelectOptionType {
    name?: string;
    title: string;
    value: string;
}
