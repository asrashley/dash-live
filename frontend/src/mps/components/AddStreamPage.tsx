import { EditStreamCard } from './EditStreamCard';
import '../../styles/mps.less';

export default function AddStreamPage() {
  return <EditStreamCard name=".add" newStream />;
}
