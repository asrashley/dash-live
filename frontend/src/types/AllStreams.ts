import { CsrfTokenCollection } from "./CsrfTokenCollection";
import { KeyMaterial } from "./KeyMaterial";
import { Stream } from "./Stream";

export interface AllStreamsJson {
    csrf_tokens: CsrfTokenCollection;
    keys: KeyMaterial[];
    streams: Stream[];
}