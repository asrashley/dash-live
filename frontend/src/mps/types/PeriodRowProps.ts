import { MpsPeriod } from "../../types/MpsPeriod";

export interface PeriodRowProps {
  className?: string;
  index: number;
  item: MpsPeriod;
}
