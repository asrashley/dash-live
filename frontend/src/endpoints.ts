import { createContext } from 'preact';
import log from 'loglevel';

import { routeMap, uiRouteMap} from '@dashlive/routemap';
import { CsrfTokenStore } from './CsrfTokenStore';
import { CsrfTokenCollection } from './types/CsrfTokenCollection';
import { AllManifests } from './types/AllManifests';
import { AllStreamsJson, AllStreamsResponse } from './types/AllStreams';
import { JWToken } from './types/JWToken';
import { MultiPeriodStream, MultiPeriodStreamJson } from './types/MultiPeriodStream';
import { DecoratedMultiPeriodStream } from "./types/DecoratedMultiPeriodStream";
import { MultiPeriodStreamSummary } from './types/MultiPeriodStreamSummary';
import { ContentRolesMap } from './types/ContentRolesMap';
import { ModifyMultiPeriodStreamJson, ModifyMultiPeriodStreamResponse } from './types/ModifyMultiPeriodStreamResponse';
import { InitialApiTokens } from './types/InitialApiTokens';
import { LoginRequest } from './types/LoginRequest';
import { LoginResponse } from './types/LoginResponse';
import { MultiPeriodStreamValidationRequest, MultiPeriodStreamValidationResponse } from './types/MpsValidation';
import { CgiOptionDescription } from './types/CgiOptionDescription';

type TokenStoreCollection = {
  files: CsrfTokenStore;
  kids: CsrfTokenStore;
  login: CsrfTokenStore;
  streams: CsrfTokenStore;
  upload: CsrfTokenStore;
};

export type ApiRequestOptions = {
  authorization?: string;
  body: RequestInit['body'];
  service: keyof TokenStoreCollection;
  signal?: AbortSignal;
  method: RequestInit['method'];
  rejectOnError?: boolean;
  query?: URLSearchParams;
};

export type GetAllStreamsProps = ApiRequestOptions & {
  withDetails: boolean
};

type RefreshAccessTokenResponse = {
  accessToken?: JWToken;
  csrfTokens: CsrfTokenCollection;
  ok: boolean;
  status: number;
};

export interface ApiRequestsProps extends Readonly<InitialApiTokens> {
  navigate: (url: string) => void;
}

export class ApiRequests {
  private csrfTokens: TokenStoreCollection;
  private accessToken: JWToken | null;
  private refreshToken: JWToken | null;
  private navigate: ApiRequestsProps['navigate'];

  constructor({csrfTokens, accessToken, refreshToken, navigate}: ApiRequestsProps) {
    this.csrfTokens = {
      files: new CsrfTokenStore(csrfTokens?.files),
      kids: new CsrfTokenStore(csrfTokens?.kids),
      login: new CsrfTokenStore(csrfTokens?.login),
      streams: new CsrfTokenStore(csrfTokens?.streams),
      upload: new CsrfTokenStore(csrfTokens?.upload),
    }
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.navigate = navigate;
  }

  getAllManifests(options: Partial<ApiRequestOptions> = {}): Promise<AllManifests> {
    return this.sendApiRequest<AllManifests>(routeMap.listManifests.url(), options);
  }

  getContentRoles(options: Partial<ApiRequestOptions> = {}): Promise<ContentRolesMap> {
    return this.sendApiRequest(routeMap.contentRoles.url(), options);
  }

  getCgiOptions(options: Partial<ApiRequestOptions> = {}) : Promise<CgiOptionDescription[]> {
    return this.sendApiRequest(routeMap.cgiOptions.url(), options);
  }

  async getUserInfo(signal: AbortSignal): Promise<LoginResponse | Response> {
    if (!this.refreshToken) {
      throw new Error('Cannot get user info without a refresh token');
    }
    const options: Partial<ApiRequestOptions> = {
      authorization: this.refreshToken.jwt,
      method: 'GET',
      rejectOnError: false,
      signal,
    };
    const response = await this.sendApiRequest<LoginResponse | Response>(routeMap.login.url(), options);
    if (response['success']) {
      this.accessToken = (response as LoginResponse).accessToken;
    }
    return response;
  }

  async loginUser(request: LoginRequest, options: Partial<GetAllStreamsProps> = {}): Promise<LoginResponse> {
    const response: LoginResponse = await this.sendApiRequest<LoginResponse>(routeMap.login.url(), {
      body: JSON.stringify(request),
      method: 'POST',
      service: 'login',
      ...options,
    });
    if (response.success) {
      if (response.accessToken) {
        this.accessToken = response.accessToken;
      }
      if (response.refreshToken) {
        this.refreshToken = response.refreshToken;
      }
    }
    return response;
  }

  async logoutUser(options: Partial<GetAllStreamsProps> = {}): Promise<Response> {
    const response: Response = await this.sendApiRequest(routeMap.login.url(), {
      method: 'DELETE',
      service: 'login',
      ...options,
    });
    this.accessToken = null;
    this.refreshToken = null;
    return response;
  }

  async getAllStreams(options: Partial<GetAllStreamsProps> = {}): Promise<AllStreamsResponse> {
    const { streams, keys } = await this.sendApiRequest<AllStreamsJson>(routeMap.listStreams.url(), options);
    return { streams, keys};
  }

  getAllMultiPeriodStreams(options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamSummary[]> {
    return this.sendApiRequest<MultiPeriodStreamSummary[]>(routeMap.listMps.url(), options);
  }

  async getMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStream> {
    const { model } = await this.sendApiRequest<MultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), options);
    return model;
  }

  async addMultiPeriodStream(
      data: DecoratedMultiPeriodStream,
      options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    const { errors, success, model } = await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.addMps.url(), {
      ...options,
      service,
      body: JSON.stringify({
        ...data,
        csrf_token
      }),
      method: 'PUT',
    });
    return { errors, success, model };
  }

  async modifyMultiPeriodStream(mps_name: string, data: DecoratedMultiPeriodStream,
     options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamJson> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      body: JSON.stringify({...data, csrf_token}),
      method: 'POST',
    });
  }

  async deleteMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<Response> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens
    );
    const query = new URLSearchParams({
      csrf_token,
    });
    return this.sendApiRequest<Response>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      query,
      method: 'DELETE',
    });
  }

  async validateMultiPeriodStream(data: MultiPeriodStreamValidationRequest,
        options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamValidationResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return this.sendApiRequest<MultiPeriodStreamValidationResponse>(routeMap.validateMps.url(), {
      ...options,
      method: 'POST',
      service: 'streams',
      body: JSON.stringify({...data, csrf_token}),
    });
  }

  private async sendApiRequest<T>(url: string, options: Partial<ApiRequestOptions>): Promise<T> {
    const { authorization, body, service, signal, method='GET',
      rejectOnError = true } = options;
    let { query } = options;
    let usedAccessToken = false;
    const headers: Headers = new Headers({
      "Content-Type": "application/json",
    });
    if (authorization !== undefined) {
      headers.set('Authorization', `Bearer ${authorization}`);
    } else {
      if (!this.accessToken && this.refreshToken) {
        await this.getAccessToken(signal);
      }
      if (this.accessToken) {
        headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
        usedAccessToken = true;
      } else if (service) {
        const token = await this.csrfTokens[service].getToken(signal);
        if (query === undefined) {
          query = new URLSearchParams({csrf_token: token});
        } else {
          query.set('csrf_token', token);
        }
      }
    }
    if (query) {
      url = `${url}?${query.toString()}`;
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const cache = 'no-cache';
    const credentials = 'same-origin';
    const mode = 'same-origin';
    url = new URL(url, document.location.origin).href;
    log.trace(`sendApiRequest(${url})`);
    let fetchResult = await fetch(url, {
      cache,
      credentials,
      mode,
      body,
      headers,
      method,
      signal,
    });
    log.trace(`url=${url} status=${fetchResult.status} usedAccessToken=${usedAccessToken}`);
    if (fetchResult.status === 401 && usedAccessToken && this.refreshToken) {
      await this.getAccessToken(signal);
      headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
      log.trace(`retrying sendApiRequest(${url})`);
      fetchResult = await fetch(url, {
          cache,
          credentials,
          mode,
          body,
          headers,
          method,
          signal,
      });
    }
    if (!fetchResult.ok) {
      if (rejectOnError) {
        throw new Error(`${ url }: ${ fetchResult.status }`);
      } else {
        return fetchResult as T;
      }
    }
    if (fetchResult.status !== 200) {
      return fetchResult as T;
    }
    const data = await fetchResult.json();
    if (signal?.aborted) {
      throw signal.reason;
    }
    if (typeof(data?.csrfTokens) === "object") {
      this.updateCsrfTokens(data.csrfTokens);
    } else if(typeof(data?.csrf_tokens) === "object") {
      this.updateCsrfTokens(data.csrf_tokens);
    } else if (typeof(data?.csrf_token) === "string") {
      this.updateCsrfTokens({[service]: data.csrf_token});
    }

    return data as T;
  }

  private updateCsrfTokens(tokens: Partial<CsrfTokenCollection>) {
    for (const [key, value] of Object.entries(tokens)) {
      if (this.csrfTokens[key] === undefined) {
        this.csrfTokens[key] = new CsrfTokenStore(value);
      } else {
        this.csrfTokens[key].setToken(value);
      }
    }
  }

  private async getAccessToken(signal: AbortSignal): Promise<JWToken> {
    if (!this.refreshToken) {
      throw new Error('Cannot request an access token without a refresh token');
    }
    const options = {
      authorization: this.refreshToken.jwt,
      rejectOnError: false,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshAccessToken.url(), options);
    const { accessToken, csrfTokens, ok, status } = data ?? {};
    if (ok === false && status === 401) {
      this.navigate(uiRouteMap.login.url());
    }
    if (!accessToken) {
      throw new Error('Failed to refresh access token');
    }
    this.accessToken = accessToken;
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
    return accessToken;
  }

  private getCsrfTokens = async (signal: AbortSignal): Promise<void> => {
    if (!this.accessToken && this.refreshToken) {
      await this.getAccessToken(signal);
    }
    if (!this.accessToken) {
      throw new Error('Cannot request CSRF tokens without an access token');
    }
    const options = {
      authorization: this.accessToken.jwt,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshCsrfTokens.url(), options);
    const { csrfTokens } = data ?? {};
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
  };
}

export const EndpointContext = createContext<ApiRequests>(null);
