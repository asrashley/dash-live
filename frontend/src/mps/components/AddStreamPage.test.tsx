import { beforeEach, describe, expect, test } from "vitest";
import { mock } from 'vitest-mock-extended';

import { renderWithProviders } from "../../test/renderWithProviders";
import AddStreamPage from "./AddStreamPage";
import { ApiRequests, EndpointContext } from "../../endpoints";
import { AllStreamsJson } from "../../types/AllStreams";
import { mediaUser } from "../../test/MockServer";

describe("AddStreamPage component", () => {
  const apiRequests = mock<ApiRequests>();

  beforeEach(() => {
    apiRequests.getAllStreams.mockImplementation(async () => {
        const streams = await import("../../test/fixtures/streams.json") as AllStreamsJson;
        return streams;
      });
  });

  test("allows a new stream to be added", async () => {
    const { asFragment, findByText, getByText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}><AddStreamPage /></EndpointContext.Provider>,
      { userInfo: mediaUser }
    );
    await findByText("Add new Multi-Period stream");
    await findByText("At least one Period is required");
    const btn = getByText("Save new stream") as HTMLButtonElement;
    expect(btn.disabled).toEqual(true);
    expect(asFragment()).toMatchSnapshot();
  });
});
