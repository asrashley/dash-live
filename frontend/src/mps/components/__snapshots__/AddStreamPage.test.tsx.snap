// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddStreamPage component > allows a new stream to be added 1`] = `
<DocumentFragment>
  <div
    class="card "
    id="edit_mps_form"
  >
    <div
      class="card-header"
    >
      <h2>
        Add new Multi-Period stream
      </h2>
    </div>
    <div
      class="card-body"
    >
      <div
        class=""
      >
        <div
          class="row mb-2 form-group "
        >
          <label
            class="col-2 col-form-label"
            for="model-name"
          >
            Name:
          </label>
          <div
            class="col-7"
          >
            <input
              class="form-control is-invalid"
              id="field-name"
              name="name"
              type="text"
            />
          </div>
          <div
            class="col-3 form-text"
          >
            Unique name for this stream
          </div>
          <div
            class="invalid-feedback"
            style="display: block;"
          >
            Name is required
          </div>
        </div>
        <div
          class="row mb-2 form-group "
        >
          <label
            class="col-2 col-form-label"
            for="model-title"
          >
            Title:
          </label>
          <div
            class="col-7"
          >
            <input
              class="form-control is-invalid"
              id="field-title"
              name="title"
              type="text"
            />
          </div>
          <div
            class="col-3 form-text"
          >
            Title for this stream
          </div>
          <div
            class="invalid-feedback"
            style="display: block;"
          >
            Title is required
          </div>
        </div>
        <div
          class="row mb-2 form-group "
        >
          <label
            class="col-2 col-form-label"
            for="model-options"
          >
            Stream Options:
          </label>
          <div
            class="col-10"
          >
            <div
              class="d-flex flex-row"
            >
              <div
                class="json bg-secondary-subtle border border-dark-subtle flex-fill me-1"
              />
              <button
                class="btn btn-primary"
              >
                Options
              </button>
            </div>
          </div>
        </div>
        <div
          class="row mb-2 form-group has-validation"
        >
          <label
            class="col-2 col-form-label"
            for="model-periods"
          >
            Periods:
          </label>
          <div
            class="col-10"
          >
            <div>
              <div
                class="period-table border"
              >
                <div
                  class="row bg-secondary text-white table-head"
                >
                  <div
                    class="col period-ordering"
                  >
                    #
                  </div>
                  <div
                    class="col period-id"
                  >
                    ID
                  </div>
                  <div
                    class="col period-stream"
                  >
                    Stream
                  </div>
                  <div
                    class="col period-start"
                  >
                    Start Time
                  </div>
                  <div
                    class="col period-duration"
                  >
                    Duration
                  </div>
                  <div
                    class="col period-tracks"
                  >
                    Tracks
                  </div>
                </div>
                <ul />
                <div
                  class="btn-toolbar"
                >
                  <button
                    class="btn btn-primary btn-sm m-2"
                  >
                    Add a Period
                  </button>
                </div>
              </div>
              <div
                class="alert alert-warning show"
                id="alert_0"
                role="alert"
              >
                At least one Period is required
              </div>
            </div>
          </div>
        </div>
        <div
          class="btn-toolbar"
        >
          <button
            class="btn btn-success m-2"
            disabled=""
          >
            Save new stream
          </button>
          <a
            class="btn btn-danger m-2"
            href="/api/multi-period-streams"
          >
            Cancel
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
